src/grain/harvest.js

```diff
@@ -0,0 +1,220 @@
+  );
+  if (!filteredSlices.length) {
+    return {
+      type: "HARVEST",
+      timestampMs,
+      version: HARVEST_VERSION_1,
+      strategy,
+      receipts: [],
+    };
+  }
+
+  const receipts: $ReadOnlyArray<GrainReceipt> = (function () {
+    switch (strategy.type) {
+      case "FAST":
+        if (strategy.version !== 1) {
+          throw new Error(`Unsupported FAST strategy: ${strategy.version}`);
+        }
+        const lastSlice = filteredSlices[filteredSlices.length - 1];
+        return computeFastReceipts(strategy.amount, lastSlice.cred);
+      case "FAIR":
+        if (strategy.version !== 1) {
+          throw new Error(`Unsupported FAIR strategy: ${strategy.version}`);
+        }
+        const totalCred = new Map();
+        for (const {cred} of filteredSlices) {
+          for (const [address, ownCred] of cred.entries()) {
+            const existingCred = totalCred.get(address) || 0;
+            totalCred.set(address, existingCred + ownCred);
+          }
+        }
+        return computeFairReceipts(strategy.amount, totalCred, earnings);
+      default:
+        throw new Error(`Unexpected type ${(strategy.type: empty)}`);
+    }
+  })();
+
+  return {
+    type: "HARVEST",
+    version: HARVEST_VERSION_1,
+    strategy,
+    receipts,
+    timestampMs,
+  };
+}
+
+/**
+ * Split a grain amount in proportion to the provided scores
+ */
+function computeFastReceipts(
+  harvestAmount: Grain,
+  cred: Map<NodeAddressT, number>
+): $ReadOnlyArray<GrainReceipt> {
+  if (harvestAmount < ZERO) {
+    throw new Error(`invalid harvestAmount: ${String(harvestAmount)}`);
+  }
+
+  const totalCred = sum(Array.from(cred.values()));
+  if (totalCred === 0) {
+    return [];
+  }
+
+  return mapToArray(cred, ([address, cred]) => ({
+    address,
+    amount: multiplyFloat(harvestAmount, cred / totalCred),
+  }));
+}
+
+/**
+ * Distribute a fixed amount of Grain to the users who were "most underpaid".
+ *
+ * We consider a user underpaid if they have recieved a smaller proportion of
+ * past earnings than their share of score. They are fairly paid if their
+ * proportion of earnings is equal to their score share, and they are overpaid
+ * if their proportion of earnings is higher than their share of the score.
+ *
+ * We start by imagining a hypothetical world, where the entire grain supply of
+ * the project (including this harvest) were distributed according to the
+ * current scores. Based on this, we can calculate the "fair" lifetime earnings
+ * for each participant. Usually, some will be "underpaid" (they recieved less
+ * than this amount) and others are "overpaid".
+ *
+ * We can sum across all users who were underpaid to find the "total
+ * underpayment". As an invariant, `totalUnderpayment = harvestAmount +
+ * totalOverpayment`, since no-one has yet been paid the harvestAmount, and
+ * beyond that every grain of overpayment to one actor is underpayment for a
+ * different actor.
```

Comment:
We can remove the comment about the invariant, since it might not hold in practice due to precision